backend/src/reset-password/reset-password.service.ts

```diff
@@ -0,0 +1,129 @@
+import { OrganizationDomainService } from 'organization-domain/organization-domain.service';
+import { Transactional } from 'typeorm-transactional-cls-hooked';
+import { CustomConfigService } from 'config/config.service';
+import { MailService } from 'mail/mail.service';
+import { ResetPasswordToken } from './entities/reset-password-token.entity';
+import { getToken, hashToken } from './reset-password.utils';
+
+@Injectable()
+export class ResetPasswordService {
+  constructor(
+    @InjectRepository(ResetPasswordToken)
+    private resetPasswordTokenRepository: Repository<ResetPasswordToken>,
+    private usersService: UsersService,
+    private organizationDomainService: OrganizationDomainService,
+    private mailService: MailService,
+    private configService: CustomConfigService,
+  ) {}
+
+  public async create(email: string): Promise<Try<string, 'USER_NOT_FOUND'>> {
+    const user = await this.usersService.findOneByEmail(email);
+
+    if (user === null) {
+      return getFailure('USER_NOT_FOUND');
+    }
+
+    const expireDate = add(new Date(), { hours: 1 });
+
+    const { hashedToken, token } = await getToken();
+    const resetPasswordToken: Omit<ResetPasswordToken, 'id' | 'userId'> = {
+      isValid: true,
+      expireDate,
+      user,
+      token: hashedToken,
+    };
+
+    await this.resetPasswordTokenRepository.insert(resetPasswordToken);
+
+    return getSuccess(token);
+  }
+
+  public async getUserByToken(
+    token: string,
+  ): Promise<Try<ResetPasswordToken, 'TOKEN_NOT_FOUND'>> {
+    const hashedToken = hashToken(token);
+    const resetToken = await this.resetPasswordTokenRepository.findOne({
+      where: {
+        token: hashedToken,
+        expireDate: MoreThanOrEqual(new Date()),
+        isValid: true,
+      },
+      relations: {
+        user: true,
+      },
+    });
+
+    if (resetToken === null) {
+      return getFailure('TOKEN_NOT_FOUND');
+    }
+
+    const { id: organizationId } =
+      this.organizationDomainService.getRequestOrganization();
+    if (resetToken.user.organizationId !== organizationId) {
+      return getFailure('TOKEN_NOT_FOUND');
+    }
+
+    return getSuccess(resetToken);
+  }
+
+  public async invalidateToken(tokenId: number) {
+    await this.resetPasswordTokenRepository.update(
+      {
+        id: tokenId,
+      },
+      {
+        isValid: false,
+      },
+    );
+  }
+
+  async initiatePasswordReset(
+    email: string,
+    isHttps: boolean,
+  ): Promise<Try<undefined, 'UNABLE_TO_CREATE_TOKEN'>> {
+    const result = await this.create(email);
+
+    if (!result.ok) {
+      return getFailure('UNABLE_TO_CREATE_TOKEN');
+    }
+
+    const token = result.data;
+    const hostname = this.configService.get('hostname');
+    const protocol = isHttps ? 'https' : 'http';
+    const { slug } = this.organizationDomainService.getRequestOrganization();
+    const resetPasswordUrl = `${protocol}://${slug}.${hostname}/account/reset-password/${token}`;
```

Comment:
Why this instead of host header (plz read)
https://www.acunetix.com/blog/articles/password-reset-poisoning/